src/main/java/duke/TaskList.java

```diff
@@ -0,0 +1,159 @@
+package duke;
+
+import duke.task.Task;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Stores taskList and the list of addCommands to be written into output file.
+ */
+public class TaskList {
+    private ArrayList<Task> tasks;
+    private ArrayList<String> addCommands;   // running these commands will always give the tasks array
+
+    /**
+     * Constructor for TaskList that initialises an empty task list.
+     */
+    public TaskList() {
+        this.tasks = new ArrayList<>();
+        this.addCommands = new ArrayList<>();
+    }
+
+    private TaskList(ArrayList<Task> tasks, ArrayList<String> addCommands) {
+        this.tasks = tasks;
+        this.addCommands = addCommands;
+    }
+
+    /**
+     * Adds a task to the task list, and writes the change to the output file.
+     *
+     * @param task Task to be added.
+     * @param command input string from user, that returns the Task's AddCommand when passed into Parser.parse().
+     * @param storage Storage to write the changes after adding.
+     * @throws IOException if an error occurs while writing to the output file.
+     */
+    public void add(Task task, String command, Storage storage) throws IOException {
+        this.tasks.add(task);
+        this.addCommands.add(command);
+
+        StringJoiner sj = new StringJoiner("\n", "", "\n");
+        this.addCommands.stream().forEach(c -> sj.add(c));
+        try {
+            storage.write(sj.toString());
+        } catch (IOException e) {
+            this.tasks.remove(task);
+            this.addCommands.remove(command);
+            throw new IOException("There was a problem writing the change to the file. duke.task.Task not added.");
+        }
+    }
+
+    public Task get(int i) throws IndexOutOfBoundsException {
```

Comment:
Perhaps its better to name the method more explicitly like getTask so there is less confusion in the future where they might be similar get methods?